onnxruntime/core/providers/cuda/cuda_graph.cc

```diff
@@ -9,15 +9,59 @@
 
 namespace onnxruntime {
 
+CudaGraphSet::~CudaGraphSet() {
+  Clear();
+}
+
+bool CudaGraphSet::IsEmpty() const {
+  return cuda_graphs_.empty();
+}
+
+void CudaGraphSet::Clear() {
+  for (auto& it : cuda_graphs_) {
+    CUDA_CALL_THROW(cudaGraphExecDestroy(it.second));
+  }
+  cuda_graphs_.clear();
+}
+
+bool CudaGraphSet::Contains(CudaGraphAnnotation_t cuda_graph_annotation_id) const {
+  return cuda_graphs_.find(cuda_graph_annotation_id) != cuda_graphs_.end();
+}
+
+void CudaGraphSet::Put(CudaGraphAnnotation_t cuda_graph_annotation_id, cudaGraphExec_t graph_exec) {
+  if (Contains(cuda_graph_annotation_id)) {
+    // This should not happen
+    throw std::runtime_error("CudaGraphSet::Put: cuda_graph_annotation_id already exists");
+  }
+  cuda_graphs_.emplace(cuda_graph_annotation_id, graph_exec);
+}
+
+bool CudaGraphSet::Get(CudaGraphAnnotation_t cuda_graph_annotation_id, cudaGraphExec_t& graph_exec) {
+  if (!Contains(cuda_graph_annotation_id)) {
+    return false;
+  }
+  graph_exec = cuda_graphs_.at(cuda_graph_annotation_id);
+  return true;
+}
+
 CUDAGraph::CUDAGraph(cudaStream_t stream) : stream_(stream) {
 }
 
 void CUDAGraph::SetStream(cudaStream_t stream) {
   stream_ = stream;
 }
 
+void CUDAGraph::SetGraphAnnotationId(CudaGraphAnnotation_t cuda_graph_annotation_id) {
+  cuda_graph_annotation_id_ = cuda_graph_annotation_id;
+}
+
 void CUDAGraph::CaptureBegin() {
-  ORT_ENFORCE(!has_graph_exec_,
+  if (!IsGraphCaptureAllowedOnRun()) {
```

Comment:
This logic shall be moved to outside. I mean, when cuda graph id is negative, we shall not create an object of CudaGraph .